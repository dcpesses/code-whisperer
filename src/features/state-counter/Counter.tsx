import { useState } from 'react';

import styles from './Counter.module.css';

export function Counter() {
  const [count, setCount] = useState(0);
  const [incrementAmount, setIncrementAmount] = useState('2');

  const increment = () => {
    setCount(count + 1);
  };
  const decrement = () => {
    setCount(count - 1);
  };
  const incrementByAmount = (val: number) => {
    setCount(count + val);
  };

  const incrementValue = Number(incrementAmount) || 0;

  return (
    <div>
      <div className={styles.row}>
        <button
          className={styles.button}
          aria-label="Decrement value"
          onClick={decrement}
        > - </button>
        <span
          className={styles.value}
          aria-label="Counter value">{count}</span>
        <button
          className={styles.button}
          aria-label="Increment value"
          onClick={increment}
        > + </button>
      </div>
      <div className={styles.row}>
        <input className={styles.textbox}
          aria-label="Set increment amount"
          value={incrementAmount}
          onChange={(e) => setIncrementAmount(e.target.value)}
        />
        <button className={styles.button}
          aria-label="Add increment amount"
          onClick={() => incrementByAmount(incrementValue)}
        >
          Add
        </button>
      </div>
    </div>
  );
}
