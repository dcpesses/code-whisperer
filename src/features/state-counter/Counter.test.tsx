import {fireEvent, render, screen} from '@testing-library/react';

import {Counter} from './Counter';

describe('Counter', () => {
  test('Should render and display state counter value', () => {
    render(<Counter />);
    const counterElement = screen.getByLabelText('Counter value', {selector: 'span'});

    expect(counterElement).toHaveTextContent('0');
    expect(counterElement).toMatchSnapshot();
  });

  test('Should increment and display updated counter value', () => {
    render(<Counter />);
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    const button = screen.getByLabelText('Increment value');

    expect(counterElement()).toHaveTextContent('0');

    fireEvent.click(button);

    expect(counterElement()).toHaveTextContent('1');
  });

  test('Should decrement and display updated counter value', () => {
    render(<Counter />);
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    const button = screen.getByLabelText('Decrement value');

    expect(counterElement()).toHaveTextContent('0');

    fireEvent.click(button);

    expect(counterElement()).toHaveTextContent('-1');
  });

  test('Should increment counter using numeric input amount', () => {
    render(<Counter />);
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    const button = screen.getByLabelText('Add increment amount');

    expect(counterElement()).toHaveTextContent('0');

    const input = screen.getByRole('textbox');
    expect(input).toHaveValue('2');
    fireEvent.change(input, {target: {value: '3'}});
    expect(input).toHaveValue('3');

    fireEvent.click(button);

    expect(counterElement()).toHaveTextContent('3');

  });
  test('Should handle invalid input values', () => {
    render(<Counter />);
    const counterElement = () => screen.getByLabelText('Counter value', {selector: 'span'});
    const button = screen.getByLabelText('Add increment amount');

    expect(counterElement()).toHaveTextContent('0');

    const input = screen.getByRole('textbox');
    fireEvent.change(input, {target: {value: 'f'}});
    fireEvent.click(button);

    expect(counterElement()).toHaveTextContent('0');

  });
});

