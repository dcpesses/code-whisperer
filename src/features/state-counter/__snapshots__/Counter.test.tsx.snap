// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Counter > Should render and display state counter value 1`] = `
<span
  aria-label="Counter value"
  class="_value_2fdd4a"
>
  0
</span>
`;
